### PR TITLE
NEXUS-4538 concurrency download test

### PR DESCRIPTION
**Pull request by velo**
Basically what this test do:

1 - download file from nexus and see how long it took
2 - start download on a new thread with limited speed (10KB/s)
3 - While limited speed download still in progress it start downloading the artifact again at full speed

Later it compares if 2 still running and if 3 did not took more then twice the time 1 took.

I ran this test on 1.9.2.2 branch as well to make sure, in fact the number of commits was high cause I went back and forward 1.9.2.2 branch.

### CODE REVIEW

File: nexus/nexus-test-harness/nexus-test-harness-launcher/src/main/java/org/sonatype/nexus/integrationtests/report/ProgressListener.java [hunk 1]
```diff
@@ -6,10 +6,10 @@
  * Public License Version 3 as published by the Free Software Foundation.
  *
  * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied
- * warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Affero General Public License Version 3
+ * warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU Affero General Public License Version 3
  * for more details.
  *
- * You should have received a copy of the GNU Affero General Public License Version 3 along with this program.  If not, see
+ * You should have received a copy of the GNU Affero General Public License Version 3 along with this program. If not, see
  * http://www.gnu.org/licenses.
```
Comment (peterlynch) [inline nexus/nexus-test-harness/nexus-test-harness-launcher/src/main/java/org/sonatype/nexus/integrationtests/report/ProgressListener.java:13]: -1 these header changes will break the build - header must match exactly. mvn license:format please

File: nexus/nexus-test-harness/nexus-test-harness-its/src/test/java/org/sonatype/nexus/integrationtests/nexus4538/Nexus4538ConcurrentDownloadIT.java [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+import org.restlet.data.Status;
+import org.sonatype.nexus.integrationtests.AbstractNexusIntegrationTest;
+import org.sonatype.nexus.tasks.descriptors.RebuildAttributesTaskDescriptor;
+import org.sonatype.nexus.test.utils.GavUtil;
+import org.sonatype.nexus.test.utils.TaskScheduleUtil;
+import org.testng.Assert;
+import org.testng.annotations.BeforeMethod;
+import org.testng.annotations.Test;
+
+public class Nexus4538ConcurrentDownloadIT
```
Comment (peterlynch) [inline nexus/nexus-test-harness/nexus-test-harness-its/src/test/java/org/sonatype/nexus/integrationtests/nexus4538/Nexus4538ConcurrentDownloadIT.java:28]: Add javadoc, describing the goal of the test, how it works and your intentions

File: nexus/nexus-test-harness/nexus-test-harness-its/src/test/java/org/sonatype/nexus/integrationtests/nexus4538/Nexus4538ConcurrentDownloadIT.java [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+    {
+        gav = GavUtil.newGav( "nexus4538", "artifact", "1.0" );
+        File f = new File( nexusWorkDir + "/storage/" + REPO_TEST_HARNESS_REPO, getRelitiveArtifactPath( gav ) );
+        populate( f, 5 );
+
+        TaskScheduleUtil.runTask( RebuildAttributesTaskDescriptor.ID );
+    }
+
+    @Test
+    public void lock()
```
Comment (peterlynch) [inline nexus/nexus-test-harness/nexus-test-harness-its/src/test/java/org/sonatype/nexus/integrationtests/nexus4538/Nexus4538ConcurrentDownloadIT.java]: I don't understand 'lock()' as a test method name. Be more descriptive like makeSureConcurrentDownloadisNotSerialized() or something. Use the method name as further documentation.

File: nexus/nexus-test-harness/nexus-test-harness-its/src/test/java/org/sonatype/nexus/integrationtests/nexus4538/Nexus4538ConcurrentDownloadIT.java [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+    private long ping( URL url )
+        throws Exception
+    {
+        long t = System.currentTimeMillis();
+        read( url, -1 );
+
+        return System.currentTimeMillis() - t;
+    }
+
+    private void read( URL url, int speedLimit )
```
Comment (peterlynch) [inline nexus/nexus-test-harness/nexus-test-harness-its/src/test/java/org/sonatype/nexus/integrationtests/nexus4538/Nexus4538ConcurrentDownloadIT.java]: Looks like a useful method to move into a utils class - someone is bound to need this functionality for other tests. When moved add javadoc

File: nexus/nexus-test-harness/nexus-test-harness-launcher/src/main/java/org/sonatype/nexus/integrationtests/RequestFacade.java [hunk 5]
```diff
@@ -57,6 +58,8 @@ import org.restlet.data.Response;
 import org.restlet.data.Status;
 import org.restlet.resource.Representation;
 import org.sonatype.nexus.test.utils.ResponseMatchers;
+import org.sonatype.nexus.test.utils.FileTestingUtils;
+import org.sonatype.nexus.test.utils.XStreamFactory;
```
Comment (peterlynch) [inline nexus/nexus-test-harness/nexus-test-harness-launcher/src/main/java/org/sonatype/nexus/integrationtests/RequestFacade.java]: Bunch of imports added but no code to go with. Doesn't that mean they should not be added?

**Comment by velo** [comment]
How can I revert changes on Nexus2497ConcurrentRepoAccessIT.java, RequestFacade.java and ProgressListener.java

git reset --hard master/origin doesn't allow me to set a path =/

File: nexus/nexus-test-harness/nexus-test-harness-its/src/test/java/org/sonatype/nexus/integrationtests/nexus4538/Nexus4538ConcurrentDownloadIT.java (hunk 4 above)
Comment (velo) [inline nexus/nexus-test-harness/nexus-test-harness-its/src/test/java/org/sonatype/nexus/integrationtests/nexus4538/Nexus4538ConcurrentDownloadIT.java]: May be move it to RequestFacade?

File: nexus/nexus-test-harness/nexus-test-harness-its/src/test/java/org/sonatype/nexus/integrationtests/nexus4538/Nexus4538ConcurrentDownloadIT.java [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+        {
+            Thread.yield();
+            Thread.sleep( 1 );
+        }
+
+        // while download is happening let's check if nexus still responsive
+        final long ping = ping( url );
+
+        // check if ping was not blocked by download
+        assertTrue( ping < ( op * 2 ), "Ping took " + ping + " original pind " + op );
```
Comment (peterlynch) [inline nexus/nexus-test-harness/nexus-test-harness-its/src/test/java/org/sonatype/nexus/integrationtests/nexus4538/Nexus4538ConcurrentDownloadIT.java]: avoid junit and testng Assert.\* except perhaps fail(), use hamcrest assertThat instead

File: nexus/nexus-test-harness/nexus-test-harness-its/src/test/java/org/sonatype/nexus/integrationtests/nexus4538/Nexus4538ConcurrentDownloadIT.java [hunk 7]
```diff
@@ -0,0 +1,168 @@
...
+
+import org.apache.commons.httpclient.HttpClient;
+import org.apache.commons.httpclient.methods.GetMethod;
+import org.apache.maven.index.artifact.Gav;
+import org.restlet.data.Status;
+import org.sonatype.nexus.integrationtests.AbstractNexusIntegrationTest;
+import org.sonatype.nexus.tasks.descriptors.RebuildAttributesTaskDescriptor;
+import org.sonatype.nexus.test.utils.GavUtil;
+import org.sonatype.nexus.test.utils.TaskScheduleUtil;
+import org.testng.Assert;
```
Comment (peterlynch) [inline nexus/nexus-test-harness/nexus-test-harness-its/src/test/java/org/sonatype/nexus/integrationtests/nexus4538/Nexus4538ConcurrentDownloadIT.java]: you should avoid using this import org.testng.Assert, even still you have it imported twice - see line 7

Comment (velo) [inline nexus/nexus-test-harness/nexus-test-harness-its/src/test/java/org/sonatype/nexus/integrationtests/nexus4538/Nexus4538ConcurrentDownloadIT.java]: import static org.testng.Assert.assertTrue != org.testng.Assert but ok

File: nexus/nexus-test-harness/nexus-test-harness-its/src/test/java/org/sonatype/nexus/integrationtests/nexus4538/Nexus4538ConcurrentDownloadIT.java (hunk 4 above)
Comment (peterlynch) [inline nexus/nexus-test-harness/nexus-test-harness-its/src/test/java/org/sonatype/nexus/integrationtests/nexus4538/Nexus4538ConcurrentDownloadIT.java]: Yeah and probably rename the method to something more descriptive,

File: nexus/nexus-test-harness/nexus-test-harness-launcher/src/main/java/org/sonatype/nexus/integrationtests/RequestFacade.java [hunk 8]
```diff
@@ -697,4 +698,65 @@ public class RequestFacade
...
+                }
+                catch ( InterruptedException e )
+                {
+                    // ignore
+                }
+            }
+        }
+
+        in.close();
+        get.releaseConnection();
```
Comment (peterlynch) [inline nexus/nexus-test-harness/nexus-test-harness-launcher/src/main/java/org/sonatype/nexus/integrationtests/RequestFacade.java]: put cleanup like this in finally block

File: nexus/nexus-test-harness/nexus-test-harness-its/src/test/java/org/sonatype/nexus/integrationtests/nexus4538/Nexus4538ConcurrentDownloadIT.java (hunk 3 above)
Comment (peterlynch) [inline nexus/nexus-test-harness/nexus-test-harness-its/src/test/java/org/sonatype/nexus/integrationtests/nexus4538/Nexus4538ConcurrentDownloadIT.java]: I forgot that this is specific to downloading the same artifact concurrently. in this case makeSureConcurrentDownloadofSameArtifactisNotSerialized() makes more sense.

**Comment by peterlynch** [comment]
other than fixing https://github.com/sonatype/nexus/commit/1170d987e15fb3025c7e4edf49af680ccc58cbb0#L0R57

+1

**Comment by bdemers** [comment]
+1
A side note, try to end javadoc descriptions with periods http://www.oracle.com/technetwork/java/javase/documentation/index-137868.html#descriptions
But to honest I don't know if that still causes problems.

**Pull request merged**